### PR TITLE
gfx942 HHS/BBS/XFP32 Grouped GEMM FreeSize logic (NN,NT,TN,TT)

### PR DESCRIPTION
**Pull request by nakajee**

### CODE REVIEW

**Comment by nakajee** [comment]
Converted from gfx942_80cu GridBased logic.
Steps to convert:
- GroupedGemm: false to true
- add "_GG_" (after Bias) in KernelNameMin and SolutionNameMin
- remove all size info
- add 2 lines with "- null" (after size info)
- change GridBased to FreeSize

**Comment by nakajee** [comment]
ActivationFuncCall is converted to true (including changes in PR #983)

**Comment by nakajee** [comment]
hipblaslt-bench verification (confirmed no issue with norm_check)

hipblaslt-bench -r f16_r --transA N --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f16_r --transA N --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f16_r --transA T --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f16_r --transA T --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA N --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA N --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA T --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r bf16_r --transA T --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA N --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA N --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA T --transB T -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info
hipblaslt-bench -r f32_r --transA T --transB N -m 1023 -n 1023 -k 1023 --batch_count 10 --compute_type xf32_r --grouped_gemm --api_method cpp --algo_method all -v --print_kernel_info

**Review by KKyang — APPROVED** [review]
All CI tests passed

**Comment by jichangjichang** [comment]
For GG, it doesn't have size mapping concept in logic yaml. You should put those yamls into gfx942 but not just for 80cu only.

**Comment by jichangjichang** [comment]
After enabling these for general gfx942, you should also need to add CI tests for these problem types for GG.

**Review by KKyang — CHANGES_REQUESTED** [review]
Sorry, I didn't see there is no gtest added.

**Comment by nakajee** [comment]
> For GG, it doesn't have size mapping concept in logic yaml. You should put those yamls into gfx942 but not just for 80cu only.

@jichangjichang , @KKyang 
I updated the files as you suggested
- added gtest for BBS/XF32 (HHS is already covered by existing gtest)
- merged HHS GG kernels into existing gfx942 GG FreeSize yamls
    - merge.py does not support FreeSize logic files. I manually merged them.
- added new BBS/XF32 GG kernels under gfx942/FreeSize (removed 80cu)

By the way, gtest failed with GSU kernel + GroupedGemm>1. I removed all GSU kernels to make it pass.

**Comment by nakajee** [comment]
gtest passed on local gfx942 node
(1) all precheckin (including new gtest)
`
./build/release/clients/staging/hipblaslt-test --gtest_filter=*pre_checkin* | tee hipblaslt_test_precheckin.log
`
(2) new tests only
`
./build/release/clients/staging/hipblaslt-test --gtest_filter=*pre_checkin_matmul_groupedgemm_bf16* | tee hipblaslt_test_precheckin_groupedgemm_bf16.log
`
`
./build/release/clients/staging/hipblaslt-test --gtest_filter=*pre_checkin_matmul_groupedgemm_xf32* | tee hipblaslt_test_precheckin_groupedgemm_xf32.log
`

**Comment by nakajee** [comment]
precheckin gtest failed because I committed old version of GG yaml (with GSU).
I reverted the previous one and re-committed the correct version (without GSU).

File: clients/gtest/matmul_gtest.yaml [hunk 1]
```diff
@@ -881,6 +881,36 @@ Tests:
...
+  M: [127, 129]
+  N: [127, 129]
+  K: [127, 129]
+  transA_transB: *transA_transB_range
+  grouped_gemm: [2]
+  alpha: 1
+  beta: [ 0.0, 2.0 ]
+  unit_check: 1
+  algo_method: [2]
+  gpu_arch: '94[0-2]'
```
Comment (jichangjichang) [inline clients/gtest/matmul_gtest.yaml]: you didn't put 942 yaml 
here should be only '942'

Comment (nakajee) [reply clients/gtest/matmul_gtest.yaml]: Thanks for your comment.
I changed to 942.

**Comment by nakajee** [comment]
@jichangjichang @KKyang 
I ran hipblaslt-test (no option) on gfx942 80cu node and confirmed all tests passed.

[==========] 35372 tests from 13 test suites ran. (1821328 ms total)
[  PASSED  ] 35372 tests.

Please let me know if you have any further comment.

**Comment by nakajee** [comment]
This is logic file update only and no change in tensilelite.
I did not run tensilelite test locally.

**Review by KKyang — APPROVED** [review]
@jichangjichang Good to go?

**Review by jichangjichang — APPROVED** [review]
lgtm, but I still have one question for step "merged HHS GG kernels into existing gfx942 GG FreeSize yamls".
gfx942 HHS already has a lot of GG solutions. Is there any special reason for that you need to merge more solutions for HHS GG?

**Comment by nakajee** [comment]
> lgtm, but I still have one question for step "merged HHS GG kernels into existing gfx942 GG FreeSize yamls". gfx942 HHS already has a lot of GG solutions. Is there any special reason for that you need to merge more solutions for HHS GG?

That is mainly because best perf kernel tends to be different between gfx942 and gfx942 80cu node.
(for example, winners on 80cu tend to have MI32 and different GLVW compared to existing gfx942 kernels)
I saw 5-10% difference when I did tuning on gfx942 80cu node.
(this is about best perf. Average was much more different)

**Comment by nakajee** [comment]
I would like to keep best perf kernels for both gfx942 and gfx942 80cu.

**Comment by jichangjichang** [comment]
> > lgtm, but I still have one question for step "merged HHS GG kernels into existing gfx942 GG FreeSize yamls". gfx942 HHS already has a lot of GG solutions. Is there any special reason for that you need to merge more solutions for HHS GG?
> 
> That is mainly because best perf kernel tends to be different between gfx942 and gfx942 80cu node. (for example, winners on 80cu tend to have MI32 and different GLVW compared to existing gfx942 kernels) I saw 5-10% difference when I did tuning on gfx942 80cu node. (this is about best perf. Average was much more different)

Got it

**Comment by nakajee** [comment]
All CI tests passed

**Pull request merged**